### PR TITLE
71 factory iterrain iresource

### LINKED ISSUE #71: Refactor ITerrain and IResource to use factories
The concrete implementation classes should be package private, and the factory should only return them through their interfaces.

### PR DESCRIPTION
**Pull request by jonathan-lindqvist**
## Summary
Terrain and resources are only created through their respective factory.

## Test Coverage
(Intellij % is bugged)
image
image
image

## Definition of Done
- [x] Completed all Tasks of the User Story
- [x] The Tasks have tests and they all pass
- [x] Overall test coverage is >= 90%
- [x] A UML class diagram of the newly implemented User Story is done and included in the README.md of the package

## Diagrams
com thebois models world resources
com thebois models world terrains

Closes #71

### CODE REVIEW

File: src/main/java/com/thebois/models/world/structures/StructureFactory.java [hunk 1]
```diff
@@ -26,7 +26,7 @@ public static void setInventory(final IInventory inventory) {
      * @param type     The type of structure to be created.
      * @param position The position of the structure.
      *
-     * @return The created structure
+     * @return The created structure.
```
Comment (martinjonsson01) [inline src/main/java/com/thebois/models/world/structures/StructureFactory.java:29]: image

File: src/test/java/com/thebois/models/world/TestWorld.java [hunk 2]
```diff
@@ -22,7 +23,11 @@ public TestWorld(final int worldSize, final ThreadLocalRandom random) {
     @Override
     protected ITerrain[][] setUpTerrain(final int worldSize, final int seed) {
         final ITerrain[][] terrainMatrix = new ITerrain[worldSize][worldSize];
-        MatrixUtils.populateElements(terrainMatrix, Grass::new);
+        for (int y = 0; y < worldSize; y++) {
+            for (int x = 0; x < worldSize; x++) {
+                terrainMatrix[y][x] = TerrainFactory.createTerrain(TerrainType.GRASS, x, y);
+            }
+        }
```
Comment (martinjonsson01) [inline src/test/java/com/thebois/models/world/TestWorld.java]: style

It was really clean before, can't you do
```java
MatrixUtils.populateElements(terrainMatrix, (x, y) -> TerrainFactory.createTerrain(TerrainType.GRASS, x, y));
```
or was there something that borked when doing it this way?

File: src/test/java/com/thebois/models/world/resources/ResourceTests.java [hunk 3]
```diff
@@ -61,7 +61,7 @@ public void resourceIsNotEqualToNull() {
     @Test
     public void resourceIsNotEqualToOtherObject() {
         // Arrange
-        final IResource resource = new Water(1, 1);
+        final IResource resource = ResourceFactory.createResource(ResourceType.WATER,1,1);
```
Comment (martinjonsson01) [inline src/test/java/com/thebois/models/world/resources/ResourceTests.java]: style

Checkstyle doesn't complain about no spaces inbetween arguments?

**Review by martinjonsson01 — APPROVED** [review]
Beautiful 👍

File: src/main/java/com/thebois/models/world/structures/StructureFactory.java (hunk 1 above)
Comment (jonathan-lindqvist) [reply src/main/java/com/thebois/models/world/structures/StructureFactory.java:29]: 🪝

File: src/test/java/com/thebois/models/world/TestWorld.java (hunk 2 above)
Comment (jonathan-lindqvist) [reply src/test/java/com/thebois/models/world/TestWorld.java]: Guess I i borked something up when I tried doing it this way, but now it works!

**Review by Matte02 — APPROVED** [review]
👍

**Pull request merged**